### PR TITLE
Escape "<" (into "<>" for stfl) feed/article title and author

### LINKED ISSUE #796: Feed title is not shown when it contains angle brackets (`<title>`)
**Newsboat version**: r2.18-240-gb462e

Config file:
```
# empty
```

Urls file:
```
http://antirez.com/rss
```

**Steps to reproduce the issue**:

1. Setup urls file, open newsboat

2. Reload feed
    - **expected**: feed title is shown as `<antirez>`
    - **actual**: feed title is empty/invisible

3. Enter feed, open first article
    - **expected**: article shows `Feed: <antirez>`
    - **actual**: article shows `Feed: `

**Other info you think is relevant**:
Executing `sqlite3 .newsboat/cache.db "select title from rss_feed where rssurl == 'http://antirez.com/rss'"` shows `<antirez>` so the title is correctly retrieved and stored in the cache.

I expect the `<antirez>` title is interpreted as a style tag.

Solution might be to encode `<` and `>` according to stfl rules.
See http://svn.clifford.at/stfl/trunk/example.stfl:
```
"a smaller-than sign: <>"
"a greater-than sign: >"
```

### PR DESCRIPTION
**Pull request by dennisschagt**
Fixes #796 

Also fixes old Newsbeuter issues:
- Fixes https://github.com/akrennmair/newsbeuter/issues/554
- Fixes https://github.com/akrennmair/newsbeuter/issues/589

@Minoru: I haven't forgotten about [your suggestion to introduce a type called StflString](https://github.com/newsboat/newsboat/pull/797#discussion_r388350769).
I'm not yet sure how that will work out so I will postpone it for now.

### CODE REVIEW

File: src/itemrenderer.cpp [hunk 1]
```diff
@@ -47,9 +47,11 @@ void prepare_header(
 	};
 
 	const std::string feedtitle = item_renderer::get_feedtitle(item);
-	add_line(feedtitle, _("Feed: "));
-	add_line(utils::utf8_to_locale(item->title()), _("Title: "));
-	add_line(utils::utf8_to_locale(item->author()), _("Author: "));
+	add_line(utils::quote_for_stfl(feedtitle), _("Feed: "));
```
Comment (Minoru) [inline src/itemrenderer.cpp]: We already have test coverage for `ItemRenderer`. Can you please add a regression test for these changes?

**Comment by Minoru** [comment]
> @Minoru: I haven't forgotten about your suggestion to introduce a type called StflString. I'm not yet sure how that will work out so I will postpone it for now.

If you don't have any immediate plans to work on that (which is fine!), I can file an issue with a more articulate description, to keep track of the idea.

File: src/itemrenderer.cpp (hunk 1 above)
Comment (dennisschagt) [reply src/itemrenderer.cpp]: Done

**Comment by dennisschagt** [comment]
> > @Minoru: I haven't forgotten about your suggestion to introduce a type called StflString. I'm not yet sure how that will work out so I will postpone it for now.
> 
> If you don't have any immediate plans to work on that (which is fine!), I can file an issue with a more articulate description, to keep track of the idea.

I have no immediate plans to pick this up so filing an issue sounds good.

File: test/itemrenderer.cpp [hunk 2]
```diff
@@ -546,3 +565,34 @@ TEST_CASE("item_renderer::get_feedtitle() returns item's feed URL "
 	const auto result = item_renderer::get_feedtitle(item);
 	REQUIRE(result == feedurl);
 }
+
+TEST_CASE("item_renderer::to_plain_text() escapes '<' for stfl",
```
Comment (Minoru) [inline test/itemrenderer.cpp]: I don't think it should behave like this. `to_plain_text()` is used to save articles in .txt files, so it shouldn't contain STFL markup. The markup should only be present in the output of `to_stfl_list()`, which is used in `ItemViewFormAction`.

(This means the other test has to be changed as well.)

Comment (dennisschagt) [reply test/itemrenderer.cpp]: Good catch. No idea why I didn't realize this myself when working on this.

The same issue already exists for the body of the articles (present before this PR, I will see if I can fix that at the same time).

**Comment by dennisschagt** [comment]
@Minoru: There is definitely room for improvement in the implementation but I first want to verify you agree with the following changes:

1. The test [`"item_renderer::to_plain_text honours `html-renderer` setting"`](https://github.com/newsboat/newsboat/blob/73a12e1c05c5ffb661609df1d609a48c300b3d50/test/itemrenderer.cpp#L365-L480) has a `"[broken]"` tag applied to it.
Do you know/remember why that was added?

    I'm guessing it is marked as such because there are html tags in the output (`"<p>Hello, world! Check out <a href='https://example.com'>our site</a>.</p>"`).
    I think it makes sense to change that output to `"Hello, world! Check out our site."`.
    Do you agree? [a]

2. This change makes the author fields of articles with an HTML author (like described in #403) look worse.
I don't think this is much of a problem, do you agree? (otherwise might make sense to wait with merging until there is a fix for that issue).

    From first article of feed referenced by #403:
    Before this PR: `"Daniel Lund&#xe9;n, Johannes Borgstr&#xf6;m, David Broman"`
    After this PR: `"<a href="http://arxiv.org/find/cs/1/au:+Lunden_D/0/1/0/all/0/1">Daniel Lund&#xe9;n</a>, <a href="http://arxiv.org/find/cs/1/au:+Borgstrom_J/0/1/0/all/0/1">Johannes Borgstr&#xf6;m</a>, <a href="http://arxiv.org/find/cs/1/au:+Broman_D/0/1/0/all/0/1">David Broman</a>"`


[a]     If you agree, I think [the code for converting `<>` to `<`](https://github.com/newsboat/newsboat/blob/48df4d1c72cf5390d250d9df40a8e15de79a99d2/src/textformatter.cpp#L216-L259) can be replaced by the [recently introduced function in RegexManager which removes tags from a string and does the `<>` to '<` conversion](https://github.com/newsboat/newsboat/blob/48df4d1c72cf5390d250d9df40a8e15de79a99d2/src/regexmanager.cpp#L81-L116).

**Comment by Minoru** [comment]
> The [issue with `item_renderer::to_plain_text()` returning STFL markup] already existed for the body of the articles (present before this PR, I will see if I can fix that at the same time).

Sorry, I didn't realize this! I do consider that to be a bug, but would prefer the fix as a separate PR. This one fixes #796 just fine, so let's just finish it off and merge it. For that, I suggest to:
1. roll back to d40e127 (the last commit that didn't have any new tests);
2. add the test "Functions used for rendering articles escape '<' into `<>` for use with stfl" as it exists in fde5a75.

The rest of the commits can become a new PR.

> Do you know/remember why [`[broken]` tag] was added?

Oh yeah, that was done by me in https://github.com/newsboat/newsboat/commit/2f46d88551f1371afde6a57ac709254e0a40bb9a. I add such tags during development in order to either ignore some tests (`./test '~[broken]'`) or to run just them (`./test '[broken]`)—very useful when you have a ton of tests failing and need to tone it down a bit to concentrate on one thing at a time. I remove those labels before submitting a PR, but this time I slipped.

This label has no purpose now, so you can remove it (perhaps linking to this comment as justification).

> I think it makes sense to change that output to "Hello, world! Check out our site.". Do you agree?

I disagree. The test that outputs HTML [uses /bin/cat as HTML renderer](https://github.com/newsboat/newsboat/blob/73a12e1c05c5ffb661609df1d609a48c300b3d50/test/itemrenderer.cpp#L460), i.e. it's expected that the output is exactly the same as the input.

It's somewhat questionable if `html-renderer` setting should apply to plain-text output, but users never reported this as a problem, so I guess it's fine.

> This change makes the author fields of articles with an HTML author (like described in #403) look worse.

Oh! Because STFL processed HTML and hid all the tags from us, and now it doesn't do that, right? I don't think we're supposed to feed HTML into STFL anyway, so it's hard for me to view this as a regression :) It's unfortunate that author names become even more unreadable, exacerbating #403, but I still intend to merge this PR. #403 is a separate problem, and can be fixed separately.

Thanks for pushing this along! :+1:

File: src/textformatter.cpp [hunk 3]
```diff
@@ -213,13 +213,60 @@ std::pair<std::string, std::size_t> TextFormatter::format_text_to_list(
 	return {format_cache, line_count};
 }
 
+std::string remove_stfl_encoding(std::string line)
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply src/textformatter.cpp]: > The rest of the commits can become a new PR.

I will take this into account in the new PR, thanks.

File: test/itemrenderer.cpp [hunk 4]
```diff
@@ -246,7 +246,7 @@ TEST_CASE("item_renderer::to_plain_text() renders text to the width specified "
 		REQUIRE(result == expected);
 	}
 
-	SECTION("stfl-encoded angle brackets are only counted as one character per `<>`") {
+	SECTION("angle brackets are only counted as one character (even if temporarily encoded as `<>`)") {
```
[the remark this answers is not in the document]
Comment (dennisschagt) [reply test/itemrenderer.cpp]: > The rest of the commits can become a new PR.

I will take this into account in the new PR, thanks.

**Comment by Minoru** [comment]
Thank you for the work, @dennisschagt!

**Pull request merged**